driver/src/price_estimation/dexag/api.rs

```diff
@@ -0,0 +1,183 @@
+#[derive(Clone, Debug)]
+pub struct Address_(pub Address);
+
+impl<'de> Deserialize<'de> for Address_ {
+    fn deserialize<D>(deserializer: D) -> Result<Address_, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        fn hex_string_to_address(string: &str) -> Result<Address> {
+            let prefix = "0x";
+            if !string.starts_with(prefix) {
+                return Err(anyhow!("does not start with {}", prefix));
+            }
+            Ok(string[2..].parse()?)
+        }
+
+        let string = String::deserialize(deserializer)?;
+        let address = hex_string_to_address(&string)
+            .with_context(|| format!("failed to parse address \"{}\"", string))
+            .map_err(D::Error::custom)?;
+        Ok(Address_(address))
+    }
+}
+
+pub enum EitherAmount {
+    From(f64),
+    To(f64),
+}
+
+// There are more dexes supported by dex.ag but we only use this one.
+pub enum Dex {
+    Ag,
+}
+
+#[derive(Debug)]
+pub struct DexagApiImpl {
+    base_url: Url,
+    client: HttpClient,
+}
+
+pub const DEFAULT_BASE_URL: &str = "https://api-v2.dex.ag";
+pub const DEFAULT_TIMEOUT: Duration = Duration::from_secs(10);
+
+impl DexagApiImpl {
+    pub fn new() -> Result<Self> {
+        Self::custom(DEFAULT_BASE_URL, DEFAULT_TIMEOUT)
+    }
+
+    pub fn custom(base_url: &str, timeout: Duration) -> Result<Self> {
+        let client = HttpClient::builder()
+            .timeout(timeout)
+            .build()
+            .context("failed to initialize HTTP client")?;
+        let base_url = base_url
+            .parse()
+            .with_context(|| format!("failed to parse url {}", base_url))?;
+        Ok(Self { base_url, client })
+    }
+}
+
+impl DexagApi for DexagApiImpl {
+    fn token_list_full(&self) -> Result<Vec<Token>> {
+        let mut url = self.base_url.clone();
+        url.set_path("token-list-full");
+        self.client
+            .get(url.to_string())
+            .context("failed to retrieve token list from dexag")?
+            .json::<Vec<Token>>()
+            .context("failed to parse token list json from dexag response")
+    }
+
+    fn price(
+        &self,
+        from: &str,
+        to: &str,
+        amount: EitherAmount,
+        dex: Dex,
+        discluded: &str,
+    ) -> Result<Price> {
+        let mut url = self.base_url.clone();
+        url.set_path("price");
+        {
+            // This is in its own block because we are supposed to drop `q`.
+            let mut q = url.query_pairs_mut();
+            q.append_pair("from", from);
+            q.append_pair("to", to);
+            match amount {
+                EitherAmount::From(amount) => q.append_pair("fromAmount", &amount.to_string()),
+                EitherAmount::To(amount) => q.append_pair("toAmount", &amount.to_string()),
+            };
+            q.append_pair(
+                "dex",
+                match dex {
+                    Dex::Ag => "ag",
```

Comment:
nit: maybe implement toStr on the enum directly?

Reply:
Was thinking about that too but its only used in this function and it could be argued that its an implementation detail which is why I slightly prefer this way.